demos/common/hfsm/az_iot_hfsm.c

```diff
@@ -0,0 +1,324 @@
+      {
+        ret = az_hfsm_transition_peer(me, idle, provisioning);
+      }
+#else
+      if(az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start))
+      {
+        LogError( ("idle: az_hfsm_post_event to _iothub_hfsm failed.") );
+        ret = az_hfsm_post_event(me, az_hfsm_errork_unknown_event);
+      }
+      else
+      {
+        ret = az_hfsm_transition_peer(me, idle, hub);
+      }
+#endif
+      break;
+
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+// AzureIoT/Provisioning
+static int32_t provisioning(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+  az_iot_hfsm_type* this_iothfsm = (az_iot_hfsm_type*)me;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+      LogInfo( ("provisioning: Entry") );
+      this_iothfsm->_retry_attempt = 0;
+      this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+      this_iothfsm->_timer_handle = az_hfsm_pal_timer_create(me);
+      break;
+
+    case AZ_HFSM_EXIT:
+      LogInfo( ("provisioning: Exit") );
+      az_hfsm_pal_timer_destroy(me, this_iothfsm->_timer_handle);
+      break;
+
+    case AZ_HFSM_TIMEOUT:
+      LogInfo( ("provisioning: Timeout") );
+      this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+      ret = az_hfsm_post_event(this_iothfsm->_provisioning_hfsm, az_hfsm_event_az_iot_start);
+      break;
+    
+    case AZ_IOT_PROVISIONING_DONE:
+      LogInfo( ("provisioning: Done") );
+      ret = az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start);
+      ret = az_hfsm_transition_peer(me, provisioning, hub);
+      break;
+
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+#endif
+
+// AzureIoT/Hub
+static int32_t hub(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+  az_iot_hfsm_type* this_iothfsm = (az_iot_hfsm_type*)me;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+      LogInfo( ("hub: Entry") );
+      this_iothfsm->_retry_attempt = 0;
+      this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+      this_iothfsm->_timer_handle = az_hfsm_pal_timer_create(me);
+      break;
+
+    case AZ_HFSM_EXIT:
+      LogInfo( ("hub: Exit") );
+      az_hfsm_pal_timer_destroy(me, this_iothfsm->_timer_handle);
+      break;
+
+    case AZ_HFSM_TIMEOUT:
+      LogInfo( ("hub: Timeout") );
+      this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+      ret = az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start);
+      break;
+
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+
+/**
+ * @brief 
+ * 
+ * @param iot_hfsm 
+ * @param provisioning_hfsm 
+ * @param hub_hfsm 
+ * @return int32_t 
+ */
+int32_t az_iot_hfsm_initialize(
+  az_iot_hfsm_type* iot_hfsm, 
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+  az_hfsm* provisioning_hfsm, 
```

Comment:
humm... make provisioning the first parameter?

Reply:
The first one is always the current state machine (iot_hfsm is the main HFSM implementing the retry logic for the two services). All parameters after that are similar to parameters passed to a constructor: the provisioning state machine and the hub state machine.